automation: feature custom headers

njmulsqb opened the pull request:
I have worked on requestor job and included the functionality to be able to add custom headers to the requests which resolves zaproxy/zaproxy#6917. It is supported in both CLI and GUI. Feel free to merge

thc202 commented:
There were still review comments in the other pull request that were not addressed.

njmulsqb commented:
> There were still review comments in the other pull request that were not addressed.

![image](https://user-images.githubusercontent.com/47940898/193900896-eb100ed0-8307-4a5c-993a-92c0ccba4edf.png)
Here's my reply on it. I have no issues in modifying it if you say so. Should I omit the added headers from both the files and just keep `headers:` in it?

njmulsqb commented on addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:
```diff
@@ -370,5 +407,13 @@ public Integer getResponseCode() {
+            return headers;
+        }
+
+        public void setHeaders(Map<?, ?> headers) {
+            this.headers = headers;
+        }
```
The headers are being handled in a Map in form of key/value pairs, so if we dont restrict headers name/key in GUI then we're expecting a map to hold a value without a key which I think is not possible.

thc202 replied:
It's possible, an empty string is a valid key like any other string.

njmulsqb replied:
How will it be expressed in YAML file then?
```
headers:
  :"value"
```
Like this? But it is invalid format I believe. Also, no header name and only value is not a valid header and does not seem even to be a valid use case

thc202 replied:
```yaml
headers:
  "":"value"
```

This was already talked in IRC, it doesn't matter if it's a valid header or not, ZAP allows to send malformed headers (to some extent, it should allow even more than it currently does). The automation jobs should not add restrictions to what the underlying functionality supports.

psiinon replied:
Well, in this case I expect the AF headers to nearly always used for submitting real headers, so I'm fine with this restriction 😁 I dont think this job needs to support every single option the underlying code supports, just the most commonly needed ones.

thc202 replied:
I'm thinking on plans to replicate issues (e.g. response splitting), so better if the job(s) don't introduce restrictions.

njmulsqb replied:
`"":"Value"` works well.

njmulsqb replied:
Do we still need the restriction to be uplifted from GUI?

njmulsqb replied:
Done

njmulsqb reviewed (commented):
It is a terrible misunderstanding, and it took me some good amount of time to realize that all my previous comments on issues were pending and not visible to you guys. https://github.com/community/community/discussions/10369 sums up the situation pretty well. Having said that, I will take care of it in future and make sure I submit my review after commenting as well.

thc202 commented on addOns/automation/src/main/resources/org/zaproxy/addon/automation/resources/requestor-max.yaml:
```diff
@@ -6,4 +6,6 @@
-        responseCode:                  # Int: An optional, expected response code against which the actual response code will be matched
\ No newline at end of file
+        responseCode:                  # Int: An optional, expected response code against which the actual response code will be matched
+        headers:                       # An optional list of headers
+            header1: "value1"
```
The former looks fine to me.

njmulsqb replied:
Done

thc202 commented on addOns/automation/src/test/java/org/zaproxy/addon/automation/jobs/RequestorJobUnitTest.java:
```diff
@@ -368,6 +368,116 @@ void shouldSendMessageWithSetValues() throws IOException {
+    }
+
+    @Test
+    void shouldNotFailIfBadHeadersList() throws IOException {
```
`shouldAcceptEmptyHeadersList` would be more clear, this is not bad/incorrect input.

njmulsqb replied:
Done

thc202 commented on addOns/automation/src/test/resources/org/zaproxy/addon/automation/resources/template-max.yaml:
```diff
@@ -64,6 +64,8 @@ jobs:
         method:                        # String: A non-empty request method, default: GET
         data:                          # String: Optional data to send in the request body, supports vars
         responseCode:                  # Int: An optional, expected response code against which the actual response code will be matched
+        headers:                       # An optional list of headers
```
Worth adding to `-min` as well.

njmulsqb replied:
I had a discussion about adding this to `-min` with @psiinon and he mentioned that its okay if we dont add it there.

thc202 replied:
What's the reason not to?

psiinon replied:
The idea was that the min template really was the minimum set of configs needed, and the headers arent needed for many requests

njmulsqb replied:
maybe because its an optional field and is not required for the minimal working of requestor job?

thc202 replied:
The optional is not the/a reason, there are many optional fields in the `-min`. I'd argue the headers are worth mentioning always but fine.

thc202 commented:
There should be a test verifying that multiple headers are sent in the expected order.

thc202 commented on addOns/automation/src/test/java/org/zaproxy/addon/automation/jobs/RequestorJobUnitTest.java:
```diff
@@ -368,6 +368,116 @@ void shouldSendMessageWithSetValues() throws IOException {
+        assertThat(progress.hasWarnings(), is(equalTo(false)));
+        ArgumentCaptor<HttpMessage> argument = ArgumentCaptor.forClass(HttpMessage.class);
+        verify(httpSender).sendAndReceive(argument.capture());
+        assertThat(argument.getValue().getRequestHeader().getHeader("header1"), is(equalTo(null)));
```
`null` means no header.

thc202 replied:
The test is called `shouldSendHeaderWithNullValue`.

njmulsqb replied:
The purpose is to send a header with no value.

thc202 replied:
No value is the empty string.

kingthorin replied:
I guess it should actually be testing that a header is removed.

njmulsqb replied:
As per my understanding of the code, ` assertThat(argument.getValue().getRequestHeader().getHeader("header1"), is(equalTo(null)));` this is matching `null` with the value of the `header1`. If you see the code in `getHeader()` it is returning the value of the header, so header is being sent but with null value or no value. The other unit test is `shouldSendHeaderWithEmptyStringValue` which sends the header with empty string as value so both are different test cases and I have ran these two cases from YAML file and they're working as expected.

thc202 replied:
> so header is being sent but with null value or no value

No, the `null` value means the header is not present in the message, if it's not in the message it's not being sent.

njmulsqb replied:
Ok, so its something to do with the message code, I'll see it. Should I remove this unit test?

njmulsqb replied:
What do we want here? I can't see the reason of putting null in YAML but definitely you have more experience so I'll go with your call on this.

thc202 replied:
It's already using "null" in the YAML, we now need to decide how the job will interpret it. IMO it could be the same as empty.

njmulsqb replied:
Yes its same as empty. I have checked it, the header is not sent with null value.

thc202 replied:
Note that there should be a test with null value (can be in `shouldSendHeadersWithSetValues` though).

njmulsqb replied:
Its the same test I wrote earlier for null value right? `header1:` like this? At the end we just will be checking that header is removed from request because of null value as you mentioned the message's behavior

kingthorin replied:
Yup, that sounds right

njmulsqb replied:
Ok I will update `shouldSendHeadersWithSetValues`

thc202 commented on addOns/automation/src/test/java/org/zaproxy/addon/automation/jobs/RequestorJobUnitTest.java:
```diff
@@ -368,6 +368,116 @@ void shouldSendMessageWithSetValues() throws IOException {
     }
 
+    @Test
+    void shouldSendHeadersWithSetValues() throws IOException {
```
This is only testing one header.

njmulsqb replied:
So should I rename it to `shouldSendHeaderWithSetValue` or should I include more headers in it. The purpose was to check if header is being sent with the request or not which is being served by sending one header.

thc202 replied:
More headers.

njmulsqb replied:
Separate unit test that sends header with same name? What should be the expected output?

thc202 replied:
It could be all in the same but also fine if it has its own test. They should all be sent as defined.

thc202 replied:
Use other type, e.g. List.

njmulsqb replied:
List of Map?

thc202 replied:
Might be a little odd in the YAML since there's only one for each list entry but should work. Or it could be a List of String, splitting the name/value in the job.

njmulsqb replied:
Might need to change lots of stuff to do this like if I make headers to be a list of string and that string may look like `header1:value1` and later on string manipulation will be required to set this in message. getter/setter of header will also be changed/removed

njmulsqb replied:
Ok I will do that but for now I need your assistance in getting things back to normal state. I messed up big time :)

njmulsqb commented:
> There should be a test verifying that multiple headers are sent in the expected order.

How does the order of headers matter?

kingthorin commented:
It "shouldn't" but apps/servers/WAFs are picky sometimes. So we should ensure we're doing what the user would expect, which IMHO would be that the message is constructed with the headers in the order presented in YAML.

thc202 commented on addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/CustomHeadersDialog.java:
```diff
@@ -0,0 +1,71 @@
+
+    @Override
+    public void save() {
+        this.cusHeader.setName(this.getStringValue(NAME_PARAM).trim());
+        this.cusHeader.setValue(this.getStringValue(VALUE_PARAM).trim());
```
Avoid trimming here to keep the value the user set.

njmulsqb replied:
Done

njmulsqb commented on addOns/automation/src/main/resources/org/zaproxy/addon/automation/resources/Messages.properties:
```diff
@@ -24,12 +24,14 @@ automation.dialog.addon.uninstalladdons = Uninstall Add-Ons:
 automation.dialog.addreq.method = Method:
 automation.dialog.addreq.data = Data:
 automation.dialog.addreq.responsecode = Response Code:
+automation.dialog.addreq.tab.customheader = Custom Headers
```
Done

njmulsqb commented:
I have made some code changes, now I can see that build is failing but this is not the case with me locally. Running `:addOns:automation:compileJava` completes successfully and all the automation tests are working fine too. Why is it failing here?

> It "shouldn't" but apps/servers/WAFs are picky sometimes. So we should ensure we're doing what the user would expect, which IMHO would be that the message is constructed with the headers in the order presented in YAML.

Alright. I have checked manually, the order is being maintained, the order of headers which is in GUI/YAML is the same to the order in which they're being sent. 

Can you hint me on how the unit test will be written for this? How the order of headers can be verified?

njmulsqb commented:
What is going on here? PR is merged? I just pushed the changes and came here to update and saw this??

psiinon commented:
No, it looks like you closed it 😦

njmulsqb commented:
I didn't. I pushed the code and I can see the message that PR is merged. I just pushed the code to this PR as usual. Anyways, how can I go back to working state?
![image](https://user-images.githubusercontent.com/47940898/194339039-ee7a5584-5246-498e-acd5-069247a4afd4.png)

I am also seeing some unrelated files in my git status 
![image](https://user-images.githubusercontent.com/47940898/194339039-ee7a5584-5246-498e-acd5-069247a4afd4.png)

psiinon commented:
Its not merged - dont delete the branch!

njmulsqb commented:
Lemme share exactly what I did (I have been doing this before without issues as well)
1. I added a unit test to verify order of headers in the request
2. I then ran `git add .` and `git commit`
3. Realized that I have added one unrelated file in the commit on which I am working on for a separate issue
4. So to remove that file from commit I ran `git reset --soft HEAD~1` and then ` git reset HEAD <file>`
5. after which I pushed using `git push --force origin HEAD`
6. came here on PR to let you guys know about the changes and saw PR is closed and merged.

psiinon commented:
So I think you just reset your branch to HEAD and pushed that - there are zero files in this PR now.
Try committing your changes and pushing again.

njmulsqb commented:
I unstaged the files that are irrelevant and then committed the rest of them, strangely it shows sign off as Simon Bennets. I have pushed the code but nothing has changed here
![image](https://user-images.githubusercontent.com/47940898/194346012-b369f3dc-e485-4450-99d1-1df2718f6932.png)

njmulsqb commented:
I have created another PR, please review. The unit test has been added to verify the order
